ffi/src/handle.rs

```diff
@@ -1,12 +1,155 @@
+    /// better performance for tree traversal operations while keeping
+    /// memory usage moderate.
+    BranchReads = 1,
+
+    /// Cache all read and write operations. This provides the best performance
+    /// but uses the most memory as it caches leaf nodes and values in addition
+    /// to branch nodes.
+    All = 2,
+}
+
+impl From<CacheReadStrategy> for firewood::manager::CacheReadStrategy {
+    fn from(strategy: CacheReadStrategy) -> Self {
+        match strategy {
+            CacheReadStrategy::WritesOnly => firewood::manager::CacheReadStrategy::WritesOnly,
+            CacheReadStrategy::BranchReads => firewood::manager::CacheReadStrategy::BranchReads,
+            CacheReadStrategy::All => firewood::manager::CacheReadStrategy::All,
+        }
+    }
+}
+
+impl From<firewood::manager::CacheReadStrategy> for CacheReadStrategy {
+    fn from(strategy: firewood::manager::CacheReadStrategy) -> Self {
+        match strategy {
+            firewood::manager::CacheReadStrategy::WritesOnly => CacheReadStrategy::WritesOnly,
+            firewood::manager::CacheReadStrategy::BranchReads => CacheReadStrategy::BranchReads,
+            firewood::manager::CacheReadStrategy::All => CacheReadStrategy::All,
+        }
+    }
+}
+
+/// Arguments for creating or opening a database. These are passed to [`fwd_open_db`]
+///
+/// [`fwd_open_db`]: crate::fwd_open_db
+#[repr(C)]
+#[derive(Debug)]
+pub struct DatabaseHandleArgs<'a> {
+    /// The path to the database file.
+    ///
+    /// This must be a valid UTF-8 string, even on Windows.
+    ///
+    /// If this is empty, an error will be returned.
+    pub path: BorrowedBytes<'a>,
+
+    /// The size of the node cache. If zero, the default size of 1,500,000 nodes
+    /// will be used.
+    pub cache_size: usize,
+
+    /// The size of the free list cache. If zero, the default size of 40,000 nodes
+    /// will be used.
+    pub free_list_cache_size: usize,
+
+    /// The maximum number of revisions to keep. If zero, the default of 128
+    /// revisions will be used.
+    ///
+    /// If this is less than 2, an error will be returned.
+    pub revisions: usize,
+
+    /// The cache read strategy to use.
+    pub strategy: CacheReadStrategy,
+
+    /// Whether to truncate the database file if it exists.
+    pub truncate: bool,
+}
+
+impl DatabaseHandleArgs<'_> {
+    const fn cache_size(&self) -> NonZeroUsize {
+        match NonZeroUsize::new(self.cache_size) {
+            Some(size) => size,
+            None => DEFAULT_CACHE_SIZE,
+        }
+    }
+
+    const fn free_list_cache_size(&self) -> NonZeroUsize {
+        match NonZeroUsize::new(self.free_list_cache_size) {
+            Some(size) => size,
+            None => DEFAULT_FREE_LIST_CACHE_SIZE,
+        }
+    }
+
+    const fn revisions(&self) -> usize {
+        if self.revisions == 0 {
+            DEFAULT_REVISIONS
+        } else {
+            // if 1, building the revision manager will fail, but that's on the caller
```

Comment:
Is it? Then why are we handling 0?

Maybe we just change the earlier if to `if self.revisions < 2`